annotations/README.md

```diff
@@ -178,6 +178,176 @@ For more details, here is an early version of the annotation language:
+
+def agg(a, b):
+  return a + b
+
+utils.help()
+utils.out("".join(functools.reduce(agg, utils.read_all(), [])))
+```
+
+For PaSh to be able to hook these aggregators correctly, we also need to add their annotations in [annotations/custom_aggregators](https://github.com/binpash/pash/tree/main/annotations/custom_aggregators).
+Below are the two annotation files named `cat.py.json` and `concat.json`. (FIXME: relative path? **Until this is fixed, prefix aggregator names with `pagg-` to avoid name clashes!**)
+The most important information in these files is (i) the aggregation command's `name`, and (ii) its treatment of inputs (both taking `["args[:]"]`).
+
+```json
+{
+    "command": "cat.py",
+    "cases":
+    [
+        {
+            "predicate": "default",
+            "class": "pure",
+            "inputs": ["args[:]"],
+            "outputs": ["stdout"]
+        }
+    ]
+}
+```
+
+```json
+{
+    "command": "concat.sh",
+    "cases":
+    [
+        {
+            "predicate": "default",
+            "class": "pure",
+            "inputs": ["args[:]"],
+            "outputs": ["stdout"]
+        }
+    ]
+}
+```
+
+
+*Step 2: Point commands to their custom aggregators*:
+Add two new annotation files in `$PASH_TOP/annotations` with names `test_one.json` and  `test_two.json`, so that they point to the right aggregator commands.
+Apart from providing the correct command `name`, the two key properties are the `class` (which should be `parallelizable_pure`) and the `rel_path` (which should point to the aggregator programs we just implemented---ideally, relative to `$PASH_TOP`).
+
+Here is the annotation for `test_one.json`, pointing to `runtime/agg/opt/concat.sh` (i.e., implying `$PASH_TOP/runtime/agg/opt/concat.sh`):
+```json
+{
+    "command": "test_one",
+    "cases":
+    [
+        {
+            "predicate": "default",
+            "class": "parallelizable_pure",
+            "inputs": ["stdin"],
+            "outputs": ["stdout"],
+            "aggregator":
+            {
+                "rel_path": "runtime/agg/opt/concat.sh",
+                "options": []
+            }
+        }
+    ]
+}
+```
+
+Here is the annotation for `test_two.json`, pointing to `runtime/agg/py/cat.py` (i.e., implying `$PASH_TOP/runtime/agg/py/cat.py`).
+The annotations also specifies that the aggregator should be called with the `-a` flag, in addition to any other flags provided to the original command by `pash`.
+
+```json
+{
+    "command": "test_one",
+    "cases":
+    [
+        {
+            "predicate": "default",
+            "class": "parallelizable_pure",
+            "inputs": ["stdin"],
+            "outputs": ["stdout"],
+            "aggregator":
+            {
+                "rel_path": "runtime/agg/opt/concat.sh",
+                "options": ["-a"]
+            }
+        }
+    ]
+}
+```
+
+**More complex aggregators**:
+Here is the `uniq` annotation that has two cases to support two different sets of flags: 
+
+```
+{
```

Comment:
I would again prefer pointing to the real annotations from the tutorial, so that we don't have the risk of lag between the README and the real annotations.

Reply:
Ah great idea!